cmd/checkConnectivity.go

```diff
@@ -0,0 +1,135 @@
+)
+
+var checkConnectivityCmd = &cobra.Command{
+	Use:          "check-connectivity",
+	Short:        "Check connectivity between the nodes and the Kubernetes API Server",
+	RunE:         checkConnectivityCmdRun,
+	SilenceUsage: true,
+}
+
+func init() {
+	checkConnectivityCmd.Flags().StringVarP(
+		&resourceGroup,
+		"resource-group", "",
+		"",
+		"Name of the resource group that contains the Kubernetes service resource.",
+	)
+	checkConnectivityCmd.Flags().StringVarP(
+		&cluster,
+		"cluster", "",
+		"",
+		"Name of the Kubernetes service resource.",
+	)
+
+	utils.AddCommonFlags(checkConnectivityCmd, &params)
+	rootCmd.AddCommand(checkConnectivityCmd)
+}
+
+func parseCheckConnectivityFlags() (*utils.Cluster, error) {
+	var cl utils.Cluster
+
+	if params.ResourceID != "" {
+		if err := utils.ParseResourceID(params.ResourceID, &cl); err != nil {
+			return nil, fmt.Errorf("failed to parse resource id: %w", err)
+		}
+	} else if params.SubscriptionID != "" && resourceGroup != "" && cluster != "" {
+		cl = utils.Cluster{
+			SubscriptionID: params.SubscriptionID,
+			ResourceGroup:  resourceGroup,
+			Cluster:        cluster,
+		}
+	} else {
+		return nil, errors.New("subscription, resource-group and cluster flags must be specified if id is not")
+	}
+
+	return &cl, nil
+}
+
+func checkConnectivityCmdRun(cmd *cobra.Command, args []string) error {
+	cl, err := parseCheckConnectivityFlags()
+	if err != nil {
+		return err
+	}
+	if params.Verbose {
+		b, _ := json.MarshalIndent(cl, "", "  ")
+		fmt.Printf("Cluster: %s\n\n", string(b))
+	}
+
+	cred, err := utils.GetCredentials()
+	if err != nil {
+		return fmt.Errorf("failed to authenticate: %w", err)
+	}
+
+	// Retrieve the VMSS information. Notice it is in the node resource group
+	// which is different from the AKS one:
+	// https://docs.microsoft.com/en-us/azure/aks/faq#why-are-two-resource-groups-created-with-aks
+	nodeRG, err := utils.GetNodeResourceGroup(cmd.Context(), cred, cl)
+	if err != nil {
+		return fmt.Errorf("couldn't get node resource group from cluster %+v: %w",
+			cl, err)
+	}
+	vmssList, err := utils.ListVMSSs(cmd.Context(), cred, cl.SubscriptionID, nodeRG)
+	if err != nil {
+		return fmt.Errorf("couldn't get VM scale set in %s/%s: %w",
+			cl.SubscriptionID, nodeRG, err)
+	}
+	if len(vmssList) != 1 {
+		return fmt.Errorf("expecting just one VMSS in %s/%s: %d found",
+			cl.SubscriptionID, nodeRG, len(vmssList))
+	}
+	vmss := vmssList[0]
+
+	// Check connectivity by executing "kubectl version" on the node. This
+	// command will try to contact the API server to get the Kubernetes version
+	// it is running. Use only the return value of the command, tough.
+	command := "kubectl --kubeconfig /var/lib/kubelet/kubeconfig version > /dev/null; echo $?"
+	vm := utils.VirtualMachineScaleSetVM{
+		SubscriptionID:    cl.SubscriptionID,
+		NodeResourceGroup: nodeRG,
+		VMScaleSet:        vmss,
+		InstanceID:        "0", // TODO: Should we iterate over all the instances of the VMSS?
```

Comment:
Should we iterate over all the instances of the VMSS?

Reply:
If we iterate over kubernetes node objects, we'll be enumerating through each VMSS instance. With that said, if for some reason there are instances that aren't registered with the apiserver, we'd miss them. I think that can be left for future work though.

Reply:
I was thinking about this discussion and [this other comment](https://github.com/Azure/kubectl-az/pull/1#discussion_r792995985). 

I think that changing the `check-connectivity` command to take the kubernetes node name is an interesting idea. Such change would align this command with the `run-command` command and they will both take: (1) Kubernetes node name or (2) VMSS instance information (subscription, node resource name, VMSS and instance ID).

Then, for the `check-connectivity` command, if the user specify `--node`, we retrieve the information of that specific VMSS instance from the Kubernetes API server and check its connectivity. Otherwise, we check the connectivity of all nodes and to do so we have two options:
1. Repeat what we do when `--node` is specified for every node the Kubernetes API server give us.
2. Retrieve the information of the VMSS of one node from the Kubernetes API server and then we go through all the instances inside that VMSS. 

The second option will cover the case when some instances are not registered with the API server but it assume that all the instances inside the same VMSS are all nodes that belong to the same cluster. Would it be correct to assume that?

I think it is something we could discuss more in details later. What do you think about requesting users to specify the `--node` in this PR and leave the iteration over all nodes for another PR?

Reply:
Let's continue this discussion on issue https://github.com/Azure/kubectl-az/issues/3.